### PR TITLE
Eliminate dead branches around typeof comparisons

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
RyuJIT will already do dead branch elimination for `typeof(X) == typeof(Y)` patterns, but we couldn't do elimination around `foo == typeof(X)`. This fixes that using whole program knowledge - if we never saw a constructed `MT` for `X`, the comparison is not going to be true. Because it needs whole program, we still scan this dead branch so in the end this doesn't save much. We can eventually do better.

I'm doing this in `SubstitutedILProvider` instead of in RyuJIT: this is because we currently only reap a small benefit from this optimization due to it only happening during compilation phase. We need to do this during scanning as well. I think I can extend it to scannig. But the extension will require the optimization to 100% guaranteed happen during codegen. We cannot rely on whether RyuJIT will feel like it. `SubstitutedILProvider` is our way to ensure the optimization will happen no matter what - the IL from the branch will be gone and RyuJIT can at most remove the comparison (we don't mind much if it's left).

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs [hunk 1]
```diff
@@ -906,9 +927,104 @@ private static bool TryExpandTypeEquality(MethodIL methodIL, byte[] body, Opcode
...
+
+            if ((flags[reader.Offset] & OpcodeFlags.BasicBlockStart) != 0)
+                return false;
+
+            if (expectGetType)
+            {
+                if (reader.ReadILOpcode() is not ILOpcode.callvirt and not ILOpcode.call)
+                    return false;
 
+                // We don't actually mind if this is not Object.GetType
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs:993]: If it is an arbitrary call, can it return a type that happens to be equal to the other type?

Or is the idea that this case will fail the `CanReferenceConstructedTypeOrCanonicalFormOfType` check below? Ie the other argument can be anything. We are just skipping the specific common patterns here to keep things simple.

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs:993]: Yes, we should be okay with any value loaded from a local or parameter. So also any value a method call could return.

We just don't have facilities to accept any value, so only a couple recognized patterns are allowed. Allowing any instance method call is less work than also checking if it's `object.GetType`.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs [hunk 2]
```diff
@@ -906,9 +927,104 @@ private static bool TryExpandTypeEquality(MethodIL methodIL, byte[] body, Opcode
...
+            // Dataflow runs on top of uninstantiated IL and we can't answer some questions there.
+            // Unfortunately this means dataflow will still see code that the rest of the system
+            // might have optimized away. It should not be a problem in practice.
+            if (knownType.ContainsSignatureVariables())
+                return false;
+
+            if (knownType.IsCanonicalDefinitionType(CanonicalFormKind.Any))
+                return false;
+
+            if (_devirtualizationManager.CanReferenceConstructedTypeOrCanonicalFormOfType(knownType))
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs:1024]: Do we need to call convert `ConvertToCanonForm` before calling `CanReferenceConstructedTypeOrCanonicalFormOfType`? Or is the type guaranteed to be normalized somehow?

**Comment by jkotas** [comment]
I have run this under debugger on this simple test:
```
using System;

static class Program
{
    static void Main(string[] args)
    {
        if (typeof(MyType) == args.GetType())
            Console.WriteLine(42);
    }
}

static class MyType
{
}
```
I would expect the substitution to trigger for it, but it is not happening. It never hits breakpoint at this line https://github.com/dotnet/runtime/pull/102248/files#diff-7c5a8ad684ce4e7f583b3bc392219bb15fc17e8400b172a2ae32d5301b7cdd0bR1027 . Is that expected?

**Comment by MichalStrehovsky** [comment]
> It never hits breakpoint at this line https://github.com/dotnet/runtime/pull/102248/files#diff-7c5a8ad684ce4e7f583b3bc392219bb15fc17e8400b172a2ae32d5301b7cdd0bR1027 . Is that expected?

Yes, you need to flip them to the more common pattern.

The problem is in the IL scanner. IL scanner only does the "downgrade result of `typeof` to necessary MethodTable" for a limited set of IL patterns as well and this one is not it. So we end up with "constructed MethodTable is needed" in the scanning phase, and this can no longer get optimized away.

https://github.com/dotnet/runtime/blob/4246ba19bd196c5f374d94e5c1fc7b21d53bd9fc/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs#L872-L897

We really need some better facilities to analyze IL in C#, but also I don't know if I want us to build a "proper" IL importer in C#.

**Comment by MichalStrehovsky** [comment]
(I plan to look into at least sharing this code between scanner and substitutions in some way.)

**Comment by jkotas** [comment]
Could you please share a program that hits this line https://github.com/dotnet/runtime/pull/102248/files#diff-7c5a8ad684ce4e7f583b3bc392219bb15fc17e8400b172a2ae32d5301b7cdd0bR1027 in the compiler? Or is this WIP and this path is not reachable yet?

**Comment by MichalStrehovsky** [comment]
> Could you please share a program that hits this line https://github.com/dotnet/runtime/pull/102248/files#diff-7c5a8ad684ce4e7f583b3bc392219bb15fc17e8400b172a2ae32d5301b7cdd0bR1027 in the compiler? Or is this WIP and this path is not reachable yet?

It's the tests that are part of this PR. We also have hits in corelib, for example:

https://github.com/dotnet/runtime/blob/4246ba19bd196c5f374d94e5c1fc7b21d53bd9fc/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/Helpers.cs#L177-L188

(The above will also be a real saving once we can do this optimization in the scanner - this is the only places that boxes DateTime and Decimal and that's a 100 kB saving on an app that uses reflection. It doesn't kick in right now, because the scanner will see we box DateTime/decimal and that destroys our opportunity to get rid of it because `DateTime`/`decimal` is referenced in typeof comparisons in other spots.)

**Comment by jkotas** [comment]
> It's the tests that are part of this PR. 

I have extracted the test into a small program:
```
using System;
using System.Runtime.CompilerServices;

static class Program
{
    static void Main(string[] args)
    {
        Type someType = GetTheType();
        if (someType == typeof(Never3))
        {
            Console.WriteLine(42);
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static Type GetTheType() => null;
}

class Never3
{
}
```
I have compiled the test in release mode (the test is under `#if !DEBUG`). Roslyn optimized out the `someType` local variable and the IL looks like this:
```
    IL_0000:  call       class [System.Runtime]System.Type Program::GetTheType()
    IL_0005:  ldtoken    MyType
    IL_000a:  call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
    IL_000f:  call       bool [System.Runtime]System.Type::op_Equality(class [System.Runtime]System.Type,
                                                                       class [System.Runtime]System.Type)
```
It fails the pattern match in `TryExpandTypeEquality_TokenOther` very early since the ldloc that the pattern match is looking for is gone. What am I missing?

**Comment by MichalStrehovsky** [comment]
Weird, I don't know how the test would pass without it. I've submitted #102374 with just the test because I don't want to switch branches locally right now.

**Comment by MichalStrehovsky** [comment]
> Weird, I don't know how the test would pass without it. I've submitted #102374 with just the test because I don't want to switch branches locally right now.

The tests are all failing in #102374 so the optimization here works. I agree that for the local case this is pretty fragile. This is another case where the expectation is that this will mostly come from a parameter in real world code. Loading it from a local was just equally cheap in the pattern match so I just allowed it. But parameter is the main use case.

**Comment by jkotas** [comment]
I have figured out one of the mysteries:

The dotnet/runtime build sets `DebugSymbols` property to true globally. `DebugSymbols` does not actually do what its name suggests. The (portable) symbols are generated regardless of whether this property is true or false. What this property actually does is that it disables C# peephole IL optimizations. The C# peephole IL optimizations break the IL patterns used by the tests in added in this PR. Setting the `DebugSymbols` to false makes the tests fail as demonstrated by https://github.com/dotnet/runtime/pull/102391 . 

The ordinary user projects out there do not set `DebugSymbols` property. I have done my quick ad-hoc test using an ordinary project and it is why it did not work for me. I will look into deleting the DebugSymbols setting so that we build and test out bits using the same settings as our users.

**Comment by jkotas** [comment]
> Yes, you need to flip them to the more common pattern.

Ok, this was the other part of the mystery. `if (t == typeof(Never))` works as expected, `if (typeof(Never) == t)` does not work as expected. The code added in this PR handles it, but the pre-existing ldtoken handling in the scanner does not.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs [hunk 3]
```diff
@@ -703,10 +703,18 @@ protected override MethodDesc ResolveVirtualMethod(MethodDesc declMethod, DefTyp
             }
 
             public override bool CanReferenceConstructedMethodTable(TypeDesc type)
-                => _constructedMethodTables.Contains(type);
+            {
+                Debug.Assert(type.NormalizeInstantiation() == type);
+                Debug.Assert(ConstructedEETypeNode.CreationAllowed(type));
+                return _constructedMethodTables.Contains(type);
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:709]: Should we also assert that we are only adding normalizations into `_constructedMethodTables` when it is populated?

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**